sdk/python/src/dagger/codegen.py

```diff
@@ -132,8 +138,48 @@ def generate(  # noqa: C901
         """,
     )
 
+    # Collect object types for all id return types.
+    # Used to replace custom scalars by objects in inputs.
+    id_map: IDMap = {}
+    for type_name, t in schema.type_map.items():
+        if not is_object_type(t):
+            continue
+        fields: dict[str, GraphQLField] = t.fields
+        for field_name, f in fields.items():
+            if field_name == "id":
+                field_type = get_named_type(f.type)
+                id_map[field_type.name] = type_name
+
+    query_type = cast(GraphQLObjectType, schema.type_map["Query"])
+    query_fields: dict[str, GraphQLField] = query_type.fields
+
+    # Collect fields under Query that receive an id argument and return
+    # an object type that also has an id field that returns the same
+    # id type as the Query field argument.
+    #
+    # Example:
+    #   `Query.directory(id: DirectoryID): Directory` matches
```

Comment:
I'm not sure if I'm 100% understanding all this, but I think there might be a corner case for certain types like `CacheVolume` where the arg name is `key` instead of `id`: https://github.com/dagger/dagger/blob/c1995ebf5a2a7bc993e9640a81890ab4e6e13579/sdk/python/src/dagger/api/gen.py#L2765

Doesn't matter right now because we don't support cache volumes as inputs or return types yet, just an FYI as I guess that might come up in the future.

Reply:
Yes, exactly. I kicked that can later for when we cross that bridge.